src/mvp/MvpServer.cpp

```diff
@@ -341,6 +345,32 @@ void MvpServer::AttackShip(std::shared_ptr<Connection> client,
     ship.attackTargetID = targetID;
 }
 
+void MvpServer::CheckTreasureIsland(int treasure_x, int treasure_y,
+                                    PlayerData &player) {
+    std::map<uint16_t, admirals::mvp::ShipData> &shipList = player.ships;
+    for (int y = treasure_y - 1; y <= treasure_y + 1; y++) {
+        if (y < 0 || y >= BOARD_SIZE) {
+            continue;
+        }
+        for (int x = treasure_x - 1; x <= treasure_x + 1; x++) {
+            if (x < 0 || x >= BOARD_SIZE) {
+                continue;
+            }
+            const uint16_t shipId = m_board[x][y];
+            if (shipId != 0) {
+                if (auto it = shipList.find(shipId); it != shipList.end()) {
+                    player.coins += 5;
+                }
+            }
+        }
+    }
+}
```

Comment:
One issue with this implementation is that is allows ships to collect coins even if they ar  just moving through the area. We should make sure the ship is not performing any actions before giving coins to their owner.  The algorithm could also be improved such that each island does not need to be checked twice (once for each player).